File: examples/cpp/my-app/src/database/TaskRepository.cpp
#include "database/TaskRepository.h"
#include "utils/Logger.h"

namespace database {

TaskRepository::TaskRepository() : db(Database::getInstance()) {}

std::shared_ptr<models::Task> TaskRepository::create(const std::string& title, const std::string& description, int projectId) {
    db.execute("INSERT INTO tasks (title, description, project_id) VALUES ('" + title + "', '" + description + "', " + std::to_string(projectId) + ")");
    return std::make_shared<models::Task>(1, title, description, projectId);
}

std::shared_ptr<models::Task> TaskRepository::findById(int id) {
    db.execute("SELECT * FROM tasks WHERE id = " + std::to_string(id));
    return nullptr;
}

std::vector<models::Task> TaskRepository::findByProject(int projectId) {
    db.execute("SELECT * FROM tasks WHERE project_id = " + std::to_string(projectId));
    return {};
}

std::vector<models::Task> TaskRepository::findByUser(int userId) {
    db.execute("SELECT * FROM tasks WHERE assignee_id = " + std::to_string(userId));
    return {};
}

std::vector<models::Task> TaskRepository::findAll() {
    db.execute("SELECT * FROM tasks");
    return {};
}

void TaskRepository::update(int id, const std::string& title, const std::string& description) {
    db.execute("UPDATE tasks SET title = '" + title + "', description = '" + description + "' WHERE id = " + std::to_string(id));
}

void TaskRepository::deleteById(int id) {
    db.execute("DELETE FROM tasks WHERE id = " + std::to_string(id));
}

void TaskRepository::save(const models::Task& task) {
    db.execute("UPDATE tasks SET title = '" + task.getTitle() + "' WHERE id = " + std::to_string(task.getId()));
}

} // namespace database

File: examples/python/my-app/src/database/task_repository.py
"""Task repository for database operations."""

from typing import List, Optional
from models.task import Task, TaskStatus
from database.database_manager import DatabaseManager
from utils.logger import Logger

class TaskRepository:
    """Repository for Task entity database operations."""
    
    def __init__(self):
        self._db = DatabaseManager.get_instance()
        self._logger = Logger.get_instance()
    
    def find_by_id(self, task_id: int) -> Optional[Task]:
        """Find task by ID."""
        self._logger.debug(f"Finding task with ID: {task_id}")
        query = "SELECT * FROM tasks WHERE id = :id"
        result = self._db.execute_query(query, {"id": task_id})
        return None
    
    def find_by_project(self, project_id: int) -> List[Task]:
        """Find all tasks for a project."""
        self._logger.debug(f"Finding tasks for project: {project_id}")
        query = "SELECT * FROM tasks WHERE project_id = :project_id"
        result = self._db.execute_query(query, {"project_id": project_id})
        return []
    
    def find_by_assignee(self, assignee_id: int) -> List[Task]:
        """Find all tasks assigned to a user."""
        self._logger.debug(f"Finding tasks for assignee: {assignee_id}")
        query = "SELECT * FROM tasks WHERE assignee_id = :assignee_id"
        result = self._db.execute_query(query, {"assignee_id": assignee_id})
        return []
    
    def find_by_status(self, status: TaskStatus) -> List[Task]:
        """Find all tasks with a specific status."""
        self._logger.debug(f"Finding tasks with status: {status.value}")
        query = "SELECT * FROM tasks WHERE status = :status"
        result = self._db.execute_query(query, {"status": status.value})
        return []
    
    def save(self, task: Task) -> Task:
        """Save task to database."""
        self._logger.info(f"Saving task: {task.title}")
        query = "INSERT INTO tasks (...) VALUES (...)"
        result = self._db.execute_query(query)
        return task
    
    def update(self, task: Task) -> Task:
        """Update existing task."""
        self._logger.info(f"Updating task: {task.id}")
        query = "UPDATE tasks SET ... WHERE id = :id"
        result = self._db.execute_query(query, {"id": task.id})
        return task
    
    def delete(self, task_id: int) -> bool:
        """Delete task by ID."""
        self._logger.warning(f"Deleting task: {task_id}")
        query = "DELETE FROM tasks WHERE id = :id"
        result = self._db.execute_query(query, {"id": task_id})
        return True
